### ISSUE TITLE
spinoff: allow tracking the current branch's tracked branch instead

### ISSUE THREAD

**Issue by glasser** [magit/magit#2438]
Now that I'm using #2414, I'm finding that my preferred setup is to have feature branches with merge (upstream) set to `origin/master` (and pushDefault set so that I push them to `origin/feature`).

I do all my actual merging feature branches back to master via GitHub's PR merge interface anyway, so the main use of the upstream is to rebase, and to see Unpushed/Unpulled sections.  In that case, I care a lot more about `origin/master` than `master`; I like being able to just run `f u` and have the Unpushed/Unpulled sections update, rather than having to switch to `master` first and pull it.

Everything about magit works great for this, except that spinoff (which I've been using extensively) won't set up this situation; it'll only set upstream to `master`.

I'd like to fix this.  Would it make sense for this to be a customize-able option?  Or a popup flag for spinoff?

(I'm also not 100% sure how to implement this in git commands... I don't think you can pass `-u UPSTREAM` to `git checkout` like you can to `git branch`, so maybe it has to be two commands?)

**Comment by tarsius** [comment]
I have considered that, but because it cannot be expressed using `branch.autoSetupMerge` ('cause the valid values are defined by Git), I haven't done that (yet?). I like the idea though, but we'll have to think hard about how this should be exposed to users.

As I have said elsewhere, I think both using `master` or `origin/master` as the upstream makes sense. I even alternate between them in different repositories. However Git itself seems to favor the latter (by default the starting-point is only automatically set as the upstream if it is a remote branch).

I tend to think we should use a Git variable. Can't really think of a good name though.

The behavior should probably be like this: if the user has opted-in, and the starting-point is `master`, and the upstream of that is `origin/master`, and `master` and `origin/master` are at the same commit, then use the latter as upstream of the new branch, otherwise use the starting-point itself as the upstream.

**Comment by glasser** [comment]
Well, that logic won't work for spinoff will it? Unless you're doing it after the update-ref...

**Comment by tarsius** [comment]
> Well, that logic won't work for spinoff will it? Unless you're doing it after the update-ref...

Actually I was mostly talking about spinoff, so I think this logic will work for that. Having to run git an additional time to accomplish this doesn't seem like a noteworthy obstacle.

**Comment by tarsius** [comment]
Please give #2444 a try.

**Comment by glasser** [comment]
#2444 looks good based on reading it, but I don't quite understand how to replace my melpa installation with a git checkout properly so I haven't tested it.

**Comment by tarsius** [comment]
If you don't mind losing your other packages and configuration while testing, then all you have to do is:

```shell
git clone git://github.com/magnars/dash.el.git /tmp/dash
git clone git://github.com/magit/magit.git /tmp/magit
cd /tmp/magit
git checkout jb/push-tweaks
make
make magit
```

**Comment by tarsius** [comment]
If merged this in #2447 now.

**Comment by glasser** [comment]
:+1:
